### ISSUE TITLE
Zero stage 1 appears to do some cpu offloading -- 2x speedup over stage 2

### ISSUE THREAD

**Issue by timohear** [microsoft/DeepSpeed#2174]
I'm experimenting with various DeepSpeed settings while finetuning gpt-j on an a6000. On one run I accidentally used:
```
    "zero_optimization": {
        "stage": 1,
        "offload_optimizer": {
            "device": "cpu", 
            "pin_memory": false
        }
    }
```
My expectation is that this should either fail directly as a non-allowed config or eventually produce an OOM error as stage 1 does on this setup when just using :
```
    "zero_optimization": {
        "stage": 1
    }
```
(the OOM is expected given the size of the model vs the VRAM)

Instead it training correctly on stage 1 with the offload_optimizer and **2x faster** than the same configuration on stage 2:
```
    "zero_optimization": {
        "stage": 2,
        "offload_optimizer": {
            "device": "cpu", 
            "pin_memory": false
        }
    }
```

Comparing the GPU usage between stage 1 and stage 2: stage 1 uses the full 48Gb of VRAM whereas stage 2 uses around 38Gb. So stage 1 is making better use of the GPU, yet it must be offloading something as it OOMs without the offload_optimizer config.

This makes no sense to me (but the 2x speedup is some comfort :-) ). What is stage 1 doing  when an offload_optimizer is specified?

System:
- Ubuntu 20.04
- a6000 (48Gb)
- Python 3.8
- DeepSpeed 0.7
- HuggingFace 4.21.0
- GPT-J

**Comment by tjruwase** [comment]
@timohear, thanks for sharing this. A while back we consolidate stage 1 and 2 to a [single file](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/zero/stage_1_and_2.py) so that cpu offload is now available for stage 1 as well. Unfortunately, we forgot to update our docs and had not previously seen interest in this feature until now. 

FYI @jeffra, who enabled this feature.

**Comment by jeffra** [comment]
Thanks @timohear, I am glad this is working well for you! Like @tjruwase mentioned we did this a while back https://github.com/microsoft/DeepSpeed/pull/1042 but I totally missed updating these docs: https://www.deepspeed.ai/docs/config-json/#optimizer-offloading. I'll update them right now though.

**Comment by timohear** [comment]
🙂 Do you know offhand why stage 1 with offload would be faster than the same config with stage 2?

**Comment by tjruwase** [comment]
If you are using gradient accumulation zero stage 2 performs gradient reduction on every backward pass which incurs extra communication. Also, can you enabling `round_robin_gradients` described [here](https://www.deepspeed.ai/docs/config-json/#zero-optimizations-for-fp16-training) to boost zero stage 2 offload performance. 
<img width="675" alt="image" src="https://user-images.githubusercontent.com/4271600/182688286-da759c4b-2ac0-4857-8344-2f332a482136.png">
